Class roster integration

rohanmaheshwari430 opened the pull request:
### Summary <!-- Required -->

This PR includes the addition of the dynamic course update script called "import-roster.ts".  It takes a list of courses that were provided in the excel sheet and gathers their data from the current class roster. Any course that caused and error was caught and added to a list of missed courses. These courses were then searched for in the previous semester class roster, where the same logic was applied. Missed courses from there were searched for in the summer roster. 

This script will save a lot of time in the future when updating the firebase with the latest courses and information.

**Edit: (11/15)**
This PR now also includes an optimized version of the import_roster script. We abstract code and by chaining promises. This helps isolate each step of the roster integration instead of performing them simultaneously (first fetch all courses after checking every semester, then transform them, then push them). 

**Edit: (2/13)**
This PR now improves upon the optimized version of the import_roster script (import_roster_v2). This was done to remove issues with side effects and nested promises. This PR also includes an optimized version of the getCourses endpoint which signficantly decreases fetch time. The improvement wad due to restructuring the hierarchy of data in the firestore. In doing so, all courses were retrieved from a single collection as opposed to different collections of different course types. The best part is that this does not influence or change how the frontend has to retreive courses for the UI. 

**Edit: (3/5)**
This PR further increases efficiency of dyanamic course fetching by implementing Promise.allSettled(). Previously, api requests were made atomically using await statements in a for loop. This caused each request to be run one at a time and then processed based on success or failure. By using Promise.allSettled(), we perform all requests concurrently, and process the results functionally by filtering by the status field (either "fulfilled" or "rejected"). Because the refactoring wasn't as direct as simply removing the for loop and replacing it with Promise.allSettled(), we introduce a few data structures such as the identifierToDA hash map to quickly grab design areas that are associated with courses that were first rejected from the current roster, but fulfilled in the next roster fetch.

 This largely improves total fetch time, and will be useful in the future when updating D@C website with the latest course roster information.

save around 5-10s of runtime. 

### Test Plan <!-- Required -->

Testing the script was trivial. I simply ran the script and had status messages printed out at every important step of the script. I also created a test collection in the firebase to check if the amount of courses added to the firebase matched the amount of courses we were looking for. There were 3 courses that could not be retrieved, but we found out that it is due to the fact that those courses are no longer available. 

<img width="437" alt="Screen Shot 2021-10-31 at 1 30 20 PM" src="https://user-images.githubusercontent.com/46412997/139595088-8ac60869-82c5-4cdb-9f25-45662419ca1a.png">

<img width="1265" alt="Screen Shot 2021-10-31 at 1 30 40 PM" src="https://user-images.githubusercontent.com/46412997/139595102-faa24c99-6385-4c91-a238-59fd509bae39.png">

rohanmaheshwari430 committed:
wrote script to grab from class roster

rohanmaheshwari430 committed:
notes

rohanmaheshwari430 committed:
added api calls to grab missed courses

rohanmaheshwari430 committed:
variable refactor

rohanmaheshwari430 committed:
lint and format

rohanmaheshwari430 committed:
disable eslint for script

rohanmaheshwari430 committed:
lint and format

katrina433 commented on server/scripts/import_roster.ts:
```diff
@@ -0,0 +1,133 @@
+      const fCourse: string[] = coursesCSV[i].tag.split(' ');
+      formattedCourses.push(fCourse);
+    }
+    // console.log(formattedCourses);
```
can remove this line before merging

katrina433 commented on server/scripts/import_roster.ts:
```diff
@@ -0,0 +1,133 @@
+      .then(async (res) => {
+        const resData: RosterResponse = (await res.data) as RosterResponse;
+        const classes = resData.data.classes as RosterCourse[];
+        // console.log(classes[0].catalogNbr);
```
this one too

katrina433 commented on server/scripts/import_roster.ts:
```diff
@@ -0,0 +1,133 @@
+      .then(async (res) => {
+        const resData: RosterResponse = (await res.data) as RosterResponse;
+        const classes = resData.data.classes as RosterCourse[];
+        // console.log(classes[0].catalogNbr);
```
and this one

katrina433 reviewed (commented):
Nice work! This will definitely make getting courses in the future a lot easier! I know you and Vishok are working on removing repetitive code, which is good. Everything else looks pretty good. Just one question: frontend course filtering by semester currently has winter as an option, but since we are not getting winter courses at all (there probably isn't any), should I just remove that filtering option?

VishokJ reviewed (commented):
Looks good! I'm working on removing some of the repetitive code in import_roster and will get that out shortly. Also, it looks like the function getPrevRosterCourses takes in the summer URL and getSummerRosterCourses takes the previous class URL. It shouldn't affect the result since we're running them back-to-back, but it's probably good to change those for clarity.

jessfeng commented on docs/static/media/remove-selection.dfa5217d.svg:
```diff
@@ -0,0 +1,9 @@
+</pattern>
+<image id="image0" width="439" height="439" xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAbcAAAG3CAYAAAA3lxvrAAAgAElEQVR4Ae3da6xt3V3X8T8UBZLWgparLyhNk6KVWAQM15hwEazxRUWgAVLSENCGAorFWHjBpfCCEBJAA5UAcgnRKAEhJl5ADXIJGDXSkCiEBAwP4Wkf+tCHcolCKZmnzzhn7XXWWnvO+Z+3McZnJztz7bXmmnOO7/j9/98z1l57nYj1vj4wIj4tIr4gIr4yIr4tIv5VRPx4RPzLiPjWiHh9RLw6Iv56RLxgvUtxZAQQQACBjQh8YkR8TkR8eUR8Q0R8X0T8+4j4sYj4z ...
+</defs>
+</svg>
```
It would help with reviewing and merging if you could add the `docs/` folder to the `.gitignore` so we're not looking through a bunch of files with automated changes.

jessfeng commented on server/endpoints/clubs/clubs_endpoints.ts:
```diff
@@ -0,0 +1,65 @@
+  const { title } = req.body;
+
+  if (title === undefined) {
+    res.send({ success: false, message: 'One or more fields is missing' });
```
Change this message to be more informative, eg. "Club name must be specified" since there's only one field that they need to input.

jessfeng commented on server/endpoints/clubs/clubs_endpoints.ts:
```diff
@@ -0,0 +1,65 @@
+  const { content } = req.body;
+
+  if (content === undefined || field === undefined || title === undefined) {
+    res.send('One or more fields is missing.');
```
Keep this `res.send()` structure consistent with the other CRUD functions you wrote, so make sure to include the `success` field and specify the `message` field.

jessfeng commented on server/scripts/import_roster.ts:
```diff
@@ -0,0 +1,133 @@
+
+const prevClassRosterURL = `https://classes.cornell.edu/api/2.0/search/classes.json?roster=${prevSem}&subject=`;
+
+const summerClassRosterURL = `https://classes.cornell.edu/api/2.0/search/classes.json?roster=${summerSem}&subject=`;
```
Nice generalization!

rohanmaheshwari430 commented:
> Nice work! This will definitely make getting courses in the future a lot easier! I know you and Vishok are working on removing repetitive code, which is good. Everything else looks pretty good. Just one question: frontend course filtering by semester currently has winter as an option, but since we are not getting winter courses at all (there probably isn't any), should I just remove that filtering option?

Thanks for the feedback! It probably is a good idea to remove the winter filter since we do not have any winter courses at the moment!

rohanmaheshwari430 commented:
> Looks good! I'm working on removing some of the repetitive code in import_roster and will get that out shortly. Also, it looks like the function getPrevRosterCourses takes in the summer URL and getSummerRosterCourses takes the previous class URL. It shouldn't affect the result since we're running them back-to-back, but it's probably good to change those for clarity.

Thanks for the feedback! I'll definitely swap the target URLs, good catch!

rohanmaheshwari430 committed:
swapped target URLS for summer and prevRoster

jessfeng commented on server/endpoints/events/events_endpoints.ts:
```diff
@@ -0,0 +1,66 @@
+  const { content } = req.body;
+
+  if (title === undefined || field === undefined || content === undefined) {
+    res.send({ success: true, message: 'one or more fields is missing' });
```
`success` should be false

jessfeng commented on server/endpoints/courses/courses_endpoints.ts:
```diff
@@ -0,0 +1,103 @@
+    courseCode === undefined ||
+    courseId === undefined
+  ) {
+    res.send('One or more fields is missing.');
```
Add `success` field

jessfeng commented on server/endpoints/majors_minors/majors_minors_endpoints.ts:
```diff
@@ -0,0 +1,79 @@
+  const { content } = req.body;
+
+  if (content === undefined || field === undefined || title === undefined) {
+    res.send('One or more fields is missing.');
```
Add `success` field

jessfeng commented on server/scripts/import_roster.ts:
```diff
@@ -0,0 +1,129 @@
+        missedCourses.push(courses[i]);
+      });
+  }
+  console.log('i finished getting courses from current roster');
```
Probably remove `console.log` stuff before merging so that it's a bit more clear, unless we really need it

jessfeng commented on server/server.ts:
```diff
@@ -41,423 +50,130 @@ export const events = db.collection('events');
-    }
-  }
-  res.send({ success: true, data: localCourses });
+  getCourses(req, res);
```
Good work with cleanup + separation!

jessfeng commented on server/scripts/import_roster.ts:
```diff
@@ -0,0 +1,129 @@
+    newCourse.set(formattedCourses[i].content);
+  }
+  console.log('i pushed the courses to firestore');
+}
```
Alrighty since Vishok's working on the redundancy, I won't comment on it lol

jessfeng reviewed (commented):
Great work! This will definitely make it easier to dynamically get courses every semester. Most changes are small things with regards to consistency and making sure not commit the 78 files of the `doc` folder to git :,)

rohanmaheshwari430 committed:
fixed naming conventions

rohanmaheshwari430 committed:
Update courses.csv

rohanmaheshwari430 committed:
added docs cache to gitignore

rohanmaheshwari430 committed:
fixed res messages and specified error

rohanmaheshwari430 committed:
format & lint

jessfeng committed:
Delete docs

jessfeng committed:
Add Vishok's optimization

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/cornell-dti/DesignAtCornell?pullRequest=44) <br/>All committers have signed the CLA.

jessfeng committed:
linting

rohanmaheshwari430 committed:
fixed bug in class roster link property

originally, the script was populating the course's roster link using only the FA21 roster. this commit provides the  correct course roster link associated with when the course was a 'hit' in the api call

rohanmaheshwari430 committed:
lint & format

rohanmaheshwari430 committed:
added design areas

rohanmaheshwari430 committed:
Merge branch 'master' into class-roster-integration

rohanmaheshwari430 committed:
updated script to fetch SP22

rohanmaheshwari430 committed:
format & linting

rohanmaheshwari430 committed:
implemented alternate script for courses retrieval

rohanmaheshwari430 committed:
replaced target function in endpoint with new function

rohanmaheshwari430 committed:
turned designAreas from string to string[]

rohanmaheshwari430 committed:
Update import_roster_v2.ts

rohanmaheshwari430 committed:
resolved nested promise issues

rohanmaheshwari430 committed:
lint & format

rohanmaheshwari430 committed:
lint

rohanmaheshwari430 committed:
format

rohanmaheshwari430 committed:
refactor courses endpoints

updated courses endpoints to reflect new db schema

rohanmaheshwari430 committed:
lint & format

rohanmaheshwari430 committed:
format

katrina433 reviewed (approved):
Looks good now! Getting courses is much faster, nice work!

rohanmaheshwari430 committed:
removed await in loop

rohanmaheshwari430 committed:
lint
